<body>
  <!---links to launch other analysis pages-->
  <nav>
    <a href='./dnaAnalysis.html' target='_blank'>DNA Analysis</a>
    <a href='./proteinAnalysis.html' target='_blank'>Protein Analysis</a>
  </nav>
  <!--creates form to submit analusis information-->
  <form>
    <!--html for sequence entry-->
    <div id=sequenceEntry>
      <h3>Sequence Entry</h3>
      <label for="sequence-entry">Enter your sequence
        (or FASTA formatted sequences) here:</label>
        <br>
      <textarea id='sequence-entry' name='sequence-entry' rows='5' cols='40'></textarea>
    </div>
    <!--html to select the format of sequence(s)-->
    <div id=formatSelect>
      <p>What format:</p>
      <input id='format-select' name='format-select' type='radio' value='single'>Single Sequence
      <br>
      <input id='format-select' name='format-select' type='radio' value='fasta'>FASTA
    </div>
    <!--html to select what analysis to run-->
    <div id=analysisSelect>
      <h3>Analysis Options</h3>
      <input type='checkbox' id = 'gcContent' name="analysis" value="gcContent">
      <label for="gcContent">GC Content</label>
      <input type='checkbox' id = 'reverseComp' name="analysis" value="reverseComp">
      <label for="reverseComp">Reverse Complement</label>
      <br>
      <input type='checkbox' id = 'transcription' name="analysis" value="transcription">
      <label for="transcription">Transcription</label>
      <input type='checkbox' id = 'translation' name="analysis" value="translation">
      <label for="translation">Translation</label>
      <br>
      <!--<input type='submit' value="Analyze" onsubmit="testOutput()">-->
      <button onclick="dnaOutput()">Analyze</button>
  </form>
  </div>
  <!--html for analysis output window-->
  <div id=analysisOutput>
    <h3>Analysis Output:</h3>
    <textarea id='output' rows='5' cols='40'></textarea>
  </div>
  <!--import Javascript functions to perform analysis-->
  <script>
  const readFASTA = text => {
    /*
    Function that allows for the reading of a FASTA formatted sequence entry.
    The function takes the text as input and returns a
    dictionary containing the header and sequence for each entry.
    */

    //creats empty object to add the header and sequence data to.
    let sequences = {};

    //loops over whole line of text, focuses on every other entry these hold
    //the header information
    for (let line = 0; line < text.split('\n').length; line += 2) {
      //crated object entry with the header as the key and the sequence as
      //the value
      sequences[text.split('\n')[line].slice(1)] = text.split('\n')[line + 1];
    }
    return sequences; //return the object
  }

  const dnaSeqCheck = sequence => {
    /*
    This function will be used to proofread the entered sequence before
    processing. The function qill return a sequence in lower-case format
    */
    //create array of the allowed bases
    const allowedChars = ['a', 'g', 'c', 't'];
    //loop over every base in the sequence
    for (var base = 0; base < sequence.length; base++) {
      //if the character is not in the allowedChars array
      if(allowedChars.indexOf(sequence[base]) === -1) {
        //throw an error telling user that the sequence contains a character
        //other than an 'a', 't', 'g', or 'c'. Breaks the loop
        throw new Error('A base in your sequence is not a nucleotide');
      }
    }
    //if no characters are wrong, return the sequence
    return (sequence);

  }

  const gcContent = sequence => {
    /*
    This function will calculate the gc content for a given sequence. Returns
    the percent gc.
    */
    //find number of 'g': (sequence.match(/g/g) || []).length;
    //find number of 'c':  (sequence.match(/c/g) || []).length;
    //add those togther and divide by length of sequence and multiply by 100
    return ((((sequence.match(/g/g) || []).length +
      (sequence.match(/c/g) || []).length) / sequence.length) * 100);
  }

  const reverseComp = sequence => {
    /*
    This function will allow for the reverse complement of an input sequence to
    be determined. Returns a string that is reverse complement.
    */
    let reverse = []; //creates empty array to add reverse base to

    //loops over bases in sequence in reverse order
    for (let i = sequence.length - 1; i >= 0; i--) {
      //if the base is an 'a', push 't' to array
      if (sequence[i] === 'a') {
        reverse.push('t');
      }
      //if the base is an 'c', push 'g' to array
      else if (sequence[i] === 'c') {
        reverse.push('g');
      }
      //if the base is an 'g', push 'c' to array
      else if (sequence[i] === 'g') {
        reverse.push('c');
      }
      //if the base is an 't', push 'a' to array
      else {
        reverse.push('a');
      }
    }
    //convert array to string, replace commas with '', and return value
    return (reverse.toString().replace(/,/g, ''));

  }

  const transcription = sequence => {
    /*
    This funcion allows for the transcription of dna to rna for the input
    sequence. Returns a string that is the rna transcript.
    */
    let trasncript = []; //create empty array to add bases to

    //loop over every base in sequence
    for (let i = 0; i < sequence.length; i++) {
      //if the base is a 't', push 'u' to array
      if (sequence[i] === 't') {
        trasncript.push('u');
      }
      //all other bases are pushed to array
      else {
        trasncript.push(sequence[i]);
      }
    }
    //convert array to string, replace commas with '', return value
    return (trasncript.toString().replace(/,/g, ''))
  }

  const translation = sequence => {
    /*
    This function will allow for the DNA sequence to be translated into the
    corresponding protein sequence. It finds the first start codon (ATG) in the
    sequence and all of the stop codons (TAA, TAG, TGA). The protein will be
    translated using the start and first stop codon. The translated protein
    is in the 1-letter code.
    */
    //create dictionary for refernce of codons and corresponding amino acid
    const codons = {
      'ttt': 'F',
      'ttc': 'F',
      'tta': 'L',
      'ttg': 'L',
      'ctt': 'L',
      'ctc': 'L',
      'cta': 'L',
      'ctg': 'L',
      'att': 'I',
      'atc': 'I',
      'ata': 'I',
      'atg': 'M',
      'gtt': 'V',
      'gtc': 'V',
      'gta': 'V',
      'gtg': 'V',
      'tct': 'S',
      'tcc': 'S',
      'tca': 'S',
      'tcg': 'S',
      'cct': 'P',
      'ccc': 'P',
      'cca': 'P',
      'ccg': 'P',
      'act': 'T',
      'acc': 'T',
      'aca': 'T',
      'acg': 'T',
      'gct': 'A',
      'gcc': 'A',
      'gca': 'A',
      'gcg': 'A',
      'tat': 'Y',
      'tac': 'Y',
      'taa': '',
      'tag': '',
      'cat': 'H',
      'cac': 'H',
      'caa': 'Q',
      'cag': 'Q',
      'aat': 'N',
      'aac': 'N',
      'aaa': 'K',
      'aag': 'K',
      'gat': 'D',
      'gac': 'D',
      'gaa': 'E',
      'gag': 'E',
      'tgt': 'C',
      'tgc': 'C',
      'tga': '',
      'tgg': 'W',
      'cgt': 'R',
      'cgc': 'R',
      'cga': 'R',
      'cgg': 'R',
      'agt': 'S',
      'agc': 'S',
      'aga': 'R',
      'agg': 'R',
      'ggt': 'G',
      'ggc': 'G',
      'gga': 'G',
      'ggg': 'G'
    };
    let startSite = sequence.indexOf('atg');//find index of first 'atg'
    let protein = [];//create empy array to push amino acids to

    //loop over entire sequence beginning at start site, increment index by
    //3 to move to the first base of the next codon
    for (let codon = startSite; codon < sequence.length; codon += 3) {
      //if the codon matches a stop codon, break out of the loop
      if (codons[sequence.slice(codon, codon + 3)] === 'taa'||
          codons[sequence.slice(codon, codon + 3)] === 'tag'||
          codons[sequence.slice(codon, codon + 3)] === 'tga') {
        break;
      }
      //otherwise, push the corresponding amino acid for the codon
      else {
        protein.push(codons[sequence.slice(codon, codon + 3)])
      }
    }
    //for some reason, a space is pushed to the end of the list, pop that
    protein.pop();
    //comert array to string, replaces commas with '', return value
    return (protein.toString().replace(/,/g, ''));
  }

  const dnaOutput = () => {
    event.preventDefault();
    //gets sequence from input textarea
    let input = document.getElementById('sequence-entry').value;
    //creates empty array to push output info to
    let outputInfo = [];
    //gets value of anaysis check boxes
    let analysisOptions = document.getElementsByName('analysis');
    //gets value of analysis radio buttons
    let inputFormat = document.getElementsByName('format-select');

    //loops over radio button options and returns which is selects
    for(let i = 0; i < inputFormat.length; i++) {
      if(inputFormat[i].checked) {
        inputFormat = inputFormat[i].value;
        break
      }
    }

    //sinlge sequence input
    if(inputFormat === 'single') {
      //checks to make sure sequence is properly formatted
      sequence = dnaSeqCheck(input);
      //loops over all check box options
      for(let i = 0; i < analysisOptions.length; i++) {
        //if option is selected
        if (analysisOptions[i].checked) {
          //performs the corresponding analysis and pushes result to output array
          if (analysisOptions[i].value === 'gcContent') {
            outputInfo.push('GC Content: ' + gcContent(sequence));
          }
          else if (analysisOptions[i].value === 'reverseComp') {
            outputInfo.push('Reverse Complement: ' + reverseComp(sequence));
          }
          else if (analysisOptions[i].value === 'transcription') {
            outputInfo.push('Transcription: ' + transcription(sequence));
          }
          else {
            outputInfo.push('Translation: ' + translation(sequence));
          }

        }
      }
      //changes output array to string and replaces commas with line breaks
      outputInfo = outputInfo.toString().replace(/,/g, '\n')


      document.getElementById("output").innerHTML = outputInfo;
    }

  //FASTA format
    else {
      //reads the fasta formatted input into an object called sequences
      let sequences = readFASTA(input);

      //loops over every entry in the object
      for (let entry in sequences) {
        //checks to make sure each sequence is properly formatted
        sequence = dnaSeqCheck(sequences[entry]);
        //pushes header to output array
        outputInfo.push(entry);
        //loops over all check box options
        for(let i = 0; i < analysisOptions.length; i++) {
          //if option is selected
          if (analysisOptions[i].checked) {
            //performs the corresponding analysis and pushes result to output array
            if (analysisOptions[i].value === 'gcContent') {
              outputInfo.push('GC Content: ' + gcContent(sequence));
            }
            else if (analysisOptions[i].value === 'reverseComp') {
              outputInfo.push('Reverse Complement: ' + reverseComp(sequence));
            }
            else if (analysisOptions[i].value === 'transcription') {
              outputInfo.push('Transcription: ' + transcription(sequence));
            }
            else {
              outputInfo.push('Translation: ' + translation(sequence));
            }

          }
        }
        //adds an extra character between each entry
        outputInfo.push('');

      }
      //changes output array to string and replaces commas with line breaks
      outputInfo = outputInfo.toString().replace(/,/g, '\n')
      document.getElementById("output").innerHTML = outputInfo;
    }


    //document.getElementById("output").innerHTML = inputFormat;
  }
  </script>
</body>

